### PR TITLE
Iss388 obsolete def type statement

### LINKED ISSUE #388: Use of DefType statement
DefTypes are evil. We need an inspection that finds them, removes them, and replaces all matching implicit types with the appropriate explicit one.

For example:

```
DefBool z

Private Function zDoSomething()
End Function
```

Should be replaced with

```
Private Function zDoSomething() As Boolean
End Function
```

### PR DESCRIPTION
**Pull request by IvenBach**
Inspection for #388.

### CODE REVIEW

File: RetailCoder.VBE/UI/Inspections/InspectionResultsControl.xaml [hunk 1]
```diff
@@ -19,8 +17,6 @@
                 <ResourceDictionary Source="../Controls/ToolBar.xaml"/>
             </ResourceDictionary.MergedDictionaries>
 
-            <converters:InvertBoolValueConverter x:Key="InvertBoolValue" />
```
Comment (Hosch250) [inline RetailCoder.VBE/UI/Inspections/InspectionResultsControl.xaml]: What's up with the deletions in this file?

File: RubberduckTests/Inspections/DefTypeStatementInspectionTests.cs [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+            var vbe = MockVbeBuilder.BuildFromSingleStandardModule(inputCode, out _);
+            using (var state = MockParser.CreateAndParse(vbe.Object))
+            {
+                var inspection = new DefTypeStatementInspection(state);
+                var inspector = InspectionsHelper.GetInspector(inspection);
+                var inspectionResults = inspector.FindIssuesAsync(state, CancellationToken.None).Result;
+
+                Assert.AreEqual(1, inspectionResults.Count());
+            }
+        }
```
Comment (Hosch250) [inline RubberduckTests/Inspections/DefTypeStatementInspectionTests.cs]: Make these tests be data-driven. You can do something like: `[TestCase("DefVar")]` and `[TestCase("DefBool")]`, etc., then define it as `DefType(string defType)` and string-interpolate the argument into the input code. Then you can have one test with all the def types.

File: RetailCoder.VBE/UI/Inspections/InspectionResultsControl.xaml (hunk 1 above)
Comment (IvenBach) [reply RetailCoder.VBE/UI/Inspections/InspectionResultsControl.xaml]: That came up as an unused converter.

File: Rubberduck.Parsing/Inspections/Resources/InspectionsUI.resx [hunk 3]
```diff
@@ -902,4 +902,13 @@ If the parameter can be null, ignore this inspection result; passing a null valu
   <data name="UnhandledOnErrorResumeNextInspectionResultFormat" xml:space="preserve">
     <value>Errors are ignored but never handled again</value>
   </data>
+  <data name="DefTypeStatementInspectionMeta" xml:space="preserve">
+    <value>The 'Def[Type]' statement is implicitly typing variables and members based on a naming/prefix convention. This stye of typing is heavily discouraged and should be avoided.</value>
```
Comment (rkapka) [inline Rubberduck.Parsing/Inspections/Resources/InspectionsUI.resx]: Typo: **stye**

Comment (rkapka) [reply Rubberduck.Parsing/Inspections/Resources/InspectionsUI.resx]: > The 'Def[Type]' statement is implicitly typing variables and members based on a naming/prefix convention.

I find this wording a bit unintelligible. How about

> Using the 'Def[Type]' statement leads to specifying types by using a prefix. This style of naming is heavily discouraged and should be avoided.

**Comment by rkapka** [comment]
@IvenBach I don't think this PR closes the issue. You created the inspection, but there's still a quick fix left to implement.

File: RetailCoder.VBE/UI/Inspections/InspectionResultsControl.xaml (hunk 1 above)
Comment (retailcoder) [reply RetailCoder.VBE/UI/Inspections/InspectionResultsControl.xaml]: @IvenBach are you 100% sure that no XAML is referring to it? XAML markup escapes ReSharper static code analysis, and a missing class will not break the build either - but things *will* go BOOM at run-time.

Comment (Vogel612) [reply RetailCoder.VBE/UI/Inspections/InspectionResultsControl.xaml]: fulltext search in the file returns no results, I'm inclined to agree that this one is unused

**Comment by Hosch250** [comment]
@rkapka is correct. However, I recommend merging this and doing the quick fix separately. We can re-open the issue manually after it is merged.

File: RubberduckTests/Inspections/DefTypeStatementInspectionTests.cs (hunk 2 above)
Comment (bclothier) [reply RubberduckTests/Inspections/DefTypeStatementInspectionTests.cs]: Your tests are all testing the functions. You should be testing the variables -- for example this:

```DefInt i
Dim i

Public Sub Foo()
  Debug.Print TypeName(i) 'Returns an integer
End Sub
```

Ditto for usages as a parameters. The inspection technically doesn't care for those stuff but you need to verify that it won't yield a false positive in cases where there is no `DefType` but implicit `Dim x` without the `As` clause.

**Comment by rkapka** [comment]
Any specific reason why this was merged, given that the PR has a WIP label?

**Comment by Vogel612** [comment]
@rkapka WIP was outdated after the last three commits and not removed. Ping @IvenBach as reminder to remove the label next time ;)

**Comment by Vogel612** [comment]
Note that PRs containing Inspections do not necessarily need to also contain a QuickFix. 
A QuickFix should explicitly declare the type on all variables that are affected by the relevant DefType statements.

In addition to that, this PR and Inspection resurfaced a bit of a missing edge-case in the Resolver that does currently not handle DefType statements **at all**.

Waiting for those two changes (while the PR owner has stepped away for a bit) would've probably taken too long, and the PR was in a mergable state anyways...

**Comment by IvenBach** [comment]
Didn't even think about the labels at all.

It will be some time before I can begin to think about any QuickFix for this, or begin to work on ithow to make

**Pull request merged**